File: sahayog_ticket/sahayog_ticket/report/sahayog_ticket_report/sahayog_ticket_report.py
import frappe


def execute(filters=None):
    """
    Sahayog Ticket Report — uses ONLY tabSahayog Ticket.
    All employee/branch/location fields are stored directly on the ticket.
    """
    columns = get_columns()
    data = get_data(filters)
    return columns, data


# ----------------------------------------
# Columns
# ----------------------------------------
def get_columns():
    return [
        {"fieldname": "ticket_id",           "label": "Ticket ID",            "fieldtype": "Link",     "options": "Sahayog Ticket", "width": 120},
        {"fieldname": "status",              "label": "Status",               "fieldtype": "Data",     "width": 100, "escape_html": 0},
        {"fieldname": "total_days",          "label": "Age",                  "fieldtype": "Int",      "width": 60},
        {"fieldname": "tat",                 "label": "TAT",                  "fieldtype": "Int",      "width": 60},
        {"fieldname": "priority",            "label": "Priority",             "fieldtype": "Data",     "width": 100},
        {"fieldname": "assigned_to_name",    "label": "Assigned To",          "fieldtype": "Data",     "width": 150},
        {"fieldname": "employee_id",         "label": "Employee ID",          "fieldtype": "Data",     "width": 100},
        {"fieldname": "employee_name",       "label": "Employee Name",        "fieldtype": "Data",     "width": 150},
        {"fieldname": "designation",         "label": "Designation",          "fieldtype": "Data",     "width": 150},
        {"fieldname": "emp_department",      "label": "Employee Department",  "fieldtype": "Data",     "width": 150},
        {"fieldname": "branch_name",         "label": "Branch",               "fieldtype": "Data",     "width": 150},
        {"fieldname": "zone",                "label": "Zone",                 "fieldtype": "Data",     "width": 120},
        {"fieldname": "region",              "label": "Region",               "fieldtype": "Data",     "width": 120},
        {"fieldname": "division",            "label": "Division",             "fieldtype": "Data",     "width": 150},
        {"fieldname": "dept_name",           "label": "Ticket Department",    "fieldtype": "Data",     "width": 150},
        {"fieldname": "ticket_type",         "label": "Ticket Type",          "fieldtype": "Link",     "options": "Ticket Type", "width": 150},
        {"fieldname": "description",         "label": "Description",          "fieldtype": "Data",     "width": 200},
        {"fieldname": "ticket_resolved_user","label": "Resolved By",          "fieldtype": "Data",     "width": 150},
        {"fieldname": "resolved_remark",     "label": "Resolved Remark",      "fieldtype": "Data",     "width": 200},
        {"fieldname": "creation",            "label": "Created On",           "fieldtype": "Datetime", "width": 180},
        {"fieldname": "ticket_resolved_on",  "label": "Resolved On",          "fieldtype": "Datetime", "width": 180},
    ]


# ----------------------------------------
# Data — Single table query, no JOINs
# ----------------------------------------
def get_data(filters=None):
    conditions = []
    params = []

    if filters:
        if filters.get("status"):
            conditions.append("st.status = %s")
            params.append(filters["status"])

        if filters.get("priority"):
            conditions.append("st.priority = %s")
            params.append(filters["priority"])

        if filters.get("division"):
            conditions.append("st.division = %s")
            params.append(filters["division"])

        if filters.get("ticket_type"):
            conditions.append("st.ticket_type = %s")
            params.append(filters["ticket_type"])

        if filters.get("employee_id"):
            conditions.append("st.employee_id = %s")
            params.append(filters["employee_id"])

        if filters.get("employee_name"):
            conditions.append("st.employee_name LIKE %s")
            params.append(f"%{filters['employee_name']}%")

        if filters.get("branch_name"):
            conditions.append("st.branch = %s")
            params.append(filters["branch_name"])

        if filters.get("department"):
            conditions.append("st.dept_name = %s")
            params.append(filters["department"])

        if filters.get("zone"):
            conditions.append("st.zone = %s")
            params.append(filters["zone"])

        if filters.get("region"):
            conditions.append("st.region = %s")
            params.append(filters["region"])

        if filters.get("assigned_to"):
            conditions.append("st.assigned_to = %s")
            params.append(filters["assigned_to"])

        from_date = filters.get("from_date")
        to_date   = filters.get("to_date")

        if from_date and to_date:
            if from_date > to_date:
                frappe.throw("From Date cannot be after To Date")
            conditions.append("st.creation BETWEEN %s AND %s")
            params.extend([f"{from_date} 00:00:00", f"{to_date} 23:59:59"])
        elif from_date:
            conditions.append("st.creation >= %s")
            params.append(f"{from_date} 00:00:00")
        elif to_date:
            conditions.append("st.creation <= %s")
            params.append(f"{to_date} 23:59:59")

    where_clause = ("WHERE " + " AND ".join(conditions)) if conditions else ""

    # ── Single query — only tabSahayog Ticket ─────────────────────────────
    query = f"""
        SELECT
            st.name                  AS ticket_id,
            st.status,
            st.priority,
            st.employee_id,
            st.employee_name,
            st.designation,
            st.emp_department,
            st.branch                AS branch_name,
            st.zone,
            st.region,
            st.division,
            st.dept_name,
            st.ticket_type,
            st.description,
            st.assigned_to_name,
            st.ticket_resolved_user,
            st.ticket_resolved_on,
            st.resolved_remark,
            st.tat,
            st.total_days,
            st.creation
        FROM `tabSahayog Ticket` st
        {where_clause}
        ORDER BY st.creation DESC
    """

    data = frappe.db.sql(query, params, as_dict=True)

    status_colors = {
        "Open":        "#2490ef",
        "In-Progress": "#f59f00",
        "Resolved":    "#28a745",
        "Closed":      "#98a6ad",
    }

    for row in data:
        status = row.get("status")
        if status in status_colors:
            color = status_colors[status]
            row["status"] = (
                f'<span style="background-color:{color}; color:white; '
                f'padding:2px 8px; border-radius:12px; font-size:12px; '
                f'font-weight:500;">{status}</span>'
            )

    return data


@frappe.whitelist()
def get_employee_department(user=None):
    if not user:
        user = frappe.session.user
    employee = frappe.db.get_value(
        "Employee",
        {"user_id": user},
        ["department"],
        as_dict=True,
    )
    if employee and employee.department:
        return employee.department
    return None


@frappe.whitelist()
def get_status_counts(department=None, from_date=None, to_date=None):
    conditions = []
    params = []
    if department:
        conditions.append("st.dept_name = %s")
        params.append(department)
    if from_date and to_date:
        conditions.append("st.creation BETWEEN %s AND %s")
        params.extend([f"{from_date} 00:00:00", f"{to_date} 23:59:59"])
    elif from_date:
        conditions.append("st.creation >= %s")
        params.append(f"{from_date} 00:00:00")
    elif to_date:
        conditions.append("st.creation <= %s")
        params.append(f"{to_date} 23:59:59")

    where = ("WHERE " + " AND ".join(conditions)) if conditions else ""
    query = f"""
        SELECT st.status, COUNT(*) AS count
        FROM `tabSahayog Ticket` st
        {where}
        GROUP BY st.status
    """
    result = frappe.db.sql(query, params, as_dict=True)
    counts = {"Open": 0, "In-Progress": 0, "Resolved": 0, "Closed": 0}
    for row in result:
        if row.status in counts:
            counts[row.status] = row.count
    return counts

File: sahayog_ticket/sahayog_ticket/report/sahayog_ticket_report/sahayog_ticket_report.js
frappe.query_reports["Sahayog Ticket Report"] = {
  onload: function (report) {
    report.page.add_inner_button("Download", function () {
      const d = new frappe.ui.Dialog({
        title: "Export Report",
        fields: [
          {
            label: "File Format",
            fieldname: "file_format",
            fieldtype: "Select",
            options: "XLSX\nCSV",
            default: "XLSX",
            reqd: 1,
          },
        ],
        primary_action_label: "Export",
        primary_action(values) {
          const format = values.file_format;
          const filters = frappe.query_report.get_filter_values();
          const data = frappe.query_report.data || [];
          const visible_idx = data.map((_, i) => i);
          const args = {
            report_name: "Sahayog Ticket Report",
            file_format_type: format === "XLSX" ? "Excel" : format,
            filters: JSON.stringify(filters),
            visible_idx: JSON.stringify(visible_idx),
            include_indentation: 0,
            include_filters: 1,
            custom_columns: "[]",
            include_hidden_columns: 0,
          };
          open_url_post("/api/method/frappe.desk.query_report.export_query", args);
          d.hide();
        },
      });
      d.show();
    });
    const isHOUser =
      frappe.session.user === "Administrator" ||
      frappe.user.has_role("HO Support Executive") ||
      frappe.user.has_role("HO Support Manager");

    if (!isHOUser) {
      frappe.call({
        method: "sahayog_ticket.sahayog_ticket.report.sahayog_ticket_report.sahayog_ticket_report.get_employee_department",
        args: { user: frappe.session.user },
        callback: function (r) {
          if (r.message) {
            frappe.query_report.set_filter_value("department", r.message);
            setTimeout(function () {
              const deptFilter = report.get_filter("department");
              if (deptFilter) {
                deptFilter.$input.prop("disabled", true);
              }
            }, 500);
          }
        },
        async: false,
      });
    }

    frappe.call({
      method: "frappe.client.get_list",
      args: {
        doctype: "Departsection",
        fields: ["dept_name"],
        limit_page_length: 0,
      },
      callback: function (r) {
        if (r.message) {
          const depts = r.message.map((d) => d.dept_name);
          const html = depts
            .map(
              (d) =>
                `<span class="department-pill" data-dept="${d}" style="cursor:pointer; margin:2px; padding:4px 12px; border-radius:16px; background:rgba(0,0,0,0.1); color:rgb(51,51,51); font-size:12px; font-weight:500;">${d}</span>`
            )
            .join("");
          function formatDate(d) {
            if (!d) return "";
            const parts = d.split("-");
            return `${parts[2]}-${parts[1]}-${parts[0]}`;
          }
          function updateDateCapsule() {
            const from = formatDate(frappe.query_report.get_filter_value("from_date"));
            const to = formatDate(frappe.query_report.get_filter_value("to_date"));
            const dateText = `Showing data from ${from} to ${to}`;
            if ($(".date-range-capsule").length) {
              $(".date-range-capsule").text(dateText);
            }
          }
          const from = formatDate(frappe.query_report.get_filter_value("from_date"));
          const to = formatDate(frappe.query_report.get_filter_value("to_date"));
          const dateHtml = `<span class="date-range-capsule" style="font-size:12px; color:#6c7681; margin-left:15px; padding:4px 12px; border-radius:16px; background:#e8f0fe; color:#1a73e8; font-weight:500;">Showing data from ${from} to ${to}</span>`;
          const capsuleHtml = `<div class="department-capsules" style="margin-bottom:10px; padding:8px 15px; background:white; border-radius:8px; border:1px solid #d1d8dd; display:flex; align-items:center; flex-wrap:wrap;"><span style="font-size:12px; color:#6c7681; margin-right:8px; font-weight:600;">Departments:</span>${html}${dateHtml}</div>`;
          const statuses = [
            { name: "Open", color: "#2490ef" },
            { name: "In-Progress", color: "#f59f00" },
            { name: "Resolved", color: "#28a745" },
            { name: "Closed", color: "#98a6ad" },
          ];
          function fetchStatusCounts(dept) {
            frappe.call({
              method: "sahayog_ticket.sahayog_ticket.report.sahayog_ticket_report.sahayog_ticket_report.get_status_counts",
              args: {
                department: dept || "",
                from_date: frappe.query_report.get_filter_value("from_date") || "",
                to_date: frappe.query_report.get_filter_value("to_date") || "",
              },
              callback: function (r) {
                if (r.message) {
                  const counts = r.message;
                  statuses.forEach((s) => {
                    const $pill = $(`.status-pill[data-status="${s.name}"]`);
                    const count = counts[s.name] || 0;
                    $pill.find(".status-count").text(count);
                  });
                }
              },
            });
          }
          const statusHtml = statuses
            .map(
              (s) =>
                `<span class="status-pill" data-status="${s.name}" style="cursor:pointer; margin:2px; padding:4px 12px; border-radius:16px; background:${s.color}; color:white; font-size:12px; font-weight:500;">${s.name} <span class="status-count" style="background:rgba(0,0,0,0.2); padding:1px 6px; border-radius:10px; font-size:11px; margin-left:4px;">0</span></span>`
            )
            .join("");
          const statusCapsuleHtml = `<div class="status-capsules" style="margin-bottom:10px; padding:8px 15px; background:white; border-radius:8px; border:1px solid #d1d8dd; display:flex; align-items:center; flex-wrap:wrap;"><span style="font-size:12px; color:#6c7681; margin-right:8px; font-weight:600;">Status:</span>${statusHtml}</div>`;
          setTimeout(function () {
            const target = $(".page-form").length
              ? $(".page-form")
              : $(".section-body").length
              ? $(".section-body")
              : $(".reports-wrapper");
            if (target.length) {
              target.before(capsuleHtml + statusCapsuleHtml);
            } else {
              $("body").find(".query-report").prepend(capsuleHtml + statusCapsuleHtml);
            }
            $(document).on("change", "[data-fieldname='from_date'], [data-fieldname='to_date']", function () {
              updateDateCapsule();
              fetchStatusCounts(frappe.query_report.get_filter_value("department"));
            });
            const activeDept = frappe.query_report.get_filter_value("department");
            fetchStatusCounts(activeDept);
            if (activeDept) {
              $(`.department-pill[data-dept="${activeDept}"]`).addClass("active-pill").css({
                background: "#006767",
                color: "white",
                "font-weight": "700",
              });
            }
            const activeStatus = frappe.query_report.get_filter_value("status");
            if (activeStatus) {
              const sColor = statuses.find((s) => s.name === activeStatus);
              $(`.status-pill[data-status="${activeStatus}"]`).addClass("active-status-pill").css({
                background: sColor ? sColor.color : "#006767",
                color: "white",
                "font-weight": "700",
                "box-shadow": "0 0 8px rgba(0,0,0,0.4)",
              });
            }
            if (!isHOUser) {
              $(".department-pill").css({
                cursor: "not-allowed",
                opacity: "0.5",
              });
            } else {
              $(document).on("click", ".department-pill", function () {
                const $pill = $(this);
                const dept = $pill.data("dept");
                const isActive = $pill.hasClass("active-pill");
                $(".department-pill").removeClass("active-pill").css({
                  background: "rgba(0,0,0,0.1)",
                  color: "rgb(51,51,51)",
                  "font-weight": "500",
                });
                if (isActive) {
                  frappe.query_report.set_filter_value("department", "");
                  fetchStatusCounts("");
                } else {
                  $pill.addClass("active-pill").css({
                    background: "#006767",
                    color: "white",
                    "font-weight": "700",
                  });
                  frappe.query_report.set_filter_value("department", dept);
                  fetchStatusCounts(dept);
                }
              });
            }
            $(document).on("click", ".status-pill", function () {
              const $pill = $(this);
              const status = $pill.data("status");
              const isActive = $pill.hasClass("active-status-pill");
              const sColor = statuses.find((s) => s.name === status);
              $(".status-pill").removeClass("active-status-pill").each(function () {
                const sName = $(this).data("status");
                const sObj = statuses.find((s) => s.name === sName);
                $(this).css({
                  background: sObj ? sObj.color : "#98a6ad",
                  color: "white",
                  "font-weight": "500",
                  "box-shadow": "none",
                });
              });
              if (isActive) {
                frappe.query_report.set_filter_value("status", "");
              } else {
                $pill.addClass("active-status-pill").css({
                  background: sColor ? sColor.color : "#006767",
                  color: "white",
                  "font-weight": "700",
                  "box-shadow": "0 0 8px rgba(0,0,0,0.4)",
                });
                frappe.query_report.set_filter_value("status", status);
              }
            });
          }, 1000);
        }
      },
    });
  },
  // Filter definitions for report data filtering
  filters: [
    {
      fieldname: "from_date",
      label: "From Date",
      fieldtype: "Date",
      default: frappe.datetime.month_start(),
      reqd: 1,
    },
    {
      fieldname: "to_date",
      label: "To Date",
      fieldtype: "Date",
      default: frappe.datetime.nowdate(),
      reqd: 1,
    },
    {
      fieldname: "status",
      label: "Status",
      fieldtype: "Select",
      options: "\nOpen\nIn-Progress\nResolved\nClosed",
      default: "",
    },
    {
      fieldname: "priority",
      label: "Priority",
      fieldtype: "Select",
      options: "\nLow\nMedium\nHigh",
      default: "",
    },
    {
      fieldname: "division",
      label: "Division",
      fieldtype: "Data",
    },
    {
      fieldname: "ticket_type",
      label: "Ticket Type",
      fieldtype: "Link",
      options: "Ticket Type",
    },
    {
      fieldname: "employee_id",
      label: "Employee ID",
      fieldtype: "Data",
    },
    {
      fieldname: "employee_name",
      label: "Employee Name",
      fieldtype: "Data",
    },
    {
      fieldname: "branch_name",
      label: "Branch Name",
      fieldtype: "Link",
      options: "Branch",
    },
    {
      fieldname: "department",
      label: "Ticket Department",
      fieldtype: "Link",
      options: "Departsection",
    },
    {
      fieldname: "zone",
      label: "Zone",
      fieldtype: "Link",
      options: "Zone",
    },
    {
      fieldname: "region",
      label: "Region",
      fieldtype: "Link",
      options: "Region",
    },
    {
      fieldname: "assigned_to",
      label: "Assigned To",
      fieldtype: "Link",
      options: "User",
    },
  ],
};
